rt/frontend/jaxrs/src/main/java/org/apache/cxf/jaxrs/provider/AbstractJAXBProvider.java

```diff
@@ -506,38 +505,41 @@ public JAXBContext getJAXBContext(Class<?> type, Type genericType) throws JAXBEx
             }
         }
 
-        synchronized (classContexts) {
-            JAXBContext context = classContexts.get(type);
-            if (context != null) {
-                return context;
-            }
+        JAXBContext context = classContexts.get(type);
+        if (context != null) {
+            return context;
         }
 
-        JAXBContext context = getPackageContext(type, genericType);
+        context = getPackageContext(type, genericType);
 
         return context != null ? context : getClassContext(type, genericType);
     }
     public JAXBContext getClassContext(Class<?> type) throws JAXBException {
         return getClassContext(type, type);
     }
     protected JAXBContext getClassContext(Class<?> type, Type genericType) throws JAXBException {
-        synchronized (classContexts) {
-            JAXBContext context = classContexts.get(type);
-            if (context == null) {
-                Class<?>[] classes;
-                if (extraClass != null) {
-                    classes = new Class[extraClass.length + 1];
-                    classes[0] = type;
-                    System.arraycopy(extraClass, 0, classes, 1, extraClass.length);
-                } else {
-                    classes = new Class[] {type};
-                }
+        final AtomicReference<JAXBException> jaxbException = new AtomicReference<>();
```

Comment:
I am wondering if there is any particular reason to use `AtomicReference`, the code block seems to be local to the function, no sharing or race conditions should be observed.

Reply:
`JAXBException extends Exception` -> should be handled inside Function or catched+rethrown in the main code for the same behaviour

Reply:
`AtomicReference` just a object holder here

Reply:
Since this is JAXB-related code, using `javax.xml.ws.Holder` would be more lightweight option I think

Reply:
PR updated